From 04713f7b76bc07ed0f24e8f2603702ef57dd42a6 Mon Sep 17 00:00:00 2001
From: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
Date: Fri, 23 May 2025 12:46:17 +0000
Subject: [PATCH] Map xinerama index to Output by output name

Currently, xinerama indices are mapped to outputs by comparing geometries.
The problem with that approach is that geometries may mismatch due to
scaling being applied.

Since version 23.1.0, Xwayland uses provided output connector names, so
we can use them to find the matching output. This approach should be less
affected by rounding issues.


(cherry picked from commit 2de458c9fa643dc56acdd03609d6dcd20598920f)

Co-authored-by: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
---
 src/utils/xcbutils.cpp     | 15 ++++++++++++++
 src/utils/xcbutils.h       |  2 ++
 src/workspace.cpp          | 41 ++++++++++++++++++++------------------
 src/workspace.h            |  1 +
 src/xwayland/drag_x.cpp    |  2 +-
 src/xwayland/selection.cpp | 17 +---------------
 src/xwayland/selection.h   |  1 -
 7 files changed, 42 insertions(+), 37 deletions(-)

diff --git a/src/utils/xcbutils.cpp b/src/utils/xcbutils.cpp
index 171158ce57d..dbf55076e65 100644
--- a/src/utils/xcbutils.cpp
+++ b/src/utils/xcbutils.cpp
@@ -672,5 +672,20 @@ QRectF nativeFloor(const QRectF &rect)
                                       nativeFloor(outputRect.width()), nativeFloor(outputRect.height())));
 }
 
+QString atomName(xcb_atom_t atom)
+{
+    xcb_connection_t *xcbConn = kwinApp()->x11Connection();
+    xcb_get_atom_name_cookie_t nameCookie = xcb_get_atom_name(xcbConn, atom);
+    xcb_get_atom_name_reply_t *nameReply = xcb_get_atom_name_reply(xcbConn, nameCookie, nullptr);
+    if (!nameReply) {
+        return QString();
+    }
+
+    const size_t length = xcb_get_atom_name_name_length(nameReply);
+    QString name = QString::fromLatin1(xcb_get_atom_name_name(nameReply), length);
+    free(nameReply);
+    return name;
+}
+
 } // namespace Xcb
 } // namespace KWin
diff --git a/src/utils/xcbutils.h b/src/utils/xcbutils.h
index 85e20c196fb..bfe6f8ae0b8 100644
--- a/src/utils/xcbutils.h
+++ b/src/utils/xcbutils.h
@@ -61,6 +61,8 @@ qreal KWIN_EXPORT nativeRound(qreal value);
  */
 QRectF KWIN_EXPORT nativeFloor(const QRectF &value);
 
+QString KWIN_EXPORT atomName(xcb_atom_t atom);
+
 // forward declaration of methods
 static void defineCursor(xcb_window_t window, xcb_cursor_t cursor);
 static void setInputFocus(xcb_window_t window, uint8_t revertTo = XCB_INPUT_FOCUS_POINTER_ROOT, xcb_timestamp_t time = xTime());
diff --git a/src/workspace.cpp b/src/workspace.cpp
index 192c5ca95eb..191a72f6c36 100644
--- a/src/workspace.cpp
+++ b/src/workspace.cpp
@@ -82,8 +82,6 @@
 #include <QDBusConnection>
 #include <QDBusPendingCall>
 #include <QMetaProperty>
-// xcb
-#include <xcb/xinerama.h>
 
 namespace KWin
 {
@@ -1185,6 +1183,16 @@ Output *Workspace::outputAt(const QPointF &pos) const
     return bestOutput;
 }
 
+Output *Workspace::findOutput(const QString &name) const
+{
+    for (Output *output : std::as_const(m_outputs)) {
+        if (output->name() == name) {
+            return output;
+        }
+    }
+    return nullptr;
+}
+
 Output *Workspace::findOutput(Output *reference, Direction direction, bool wrapAround) const
 {
     QList<Output *> relevantOutputs;
@@ -2582,31 +2590,26 @@ Output *Workspace::xineramaIndexToOutput(int index) const
         return nullptr;
     }
 
-    const UniqueCPtr<xcb_xinerama_is_active_reply_t> active{xcb_xinerama_is_active_reply(connection, xcb_xinerama_is_active(connection), nullptr)};
-    if (!active || !active->state) {
+    xcb_randr_get_monitors_cookie_t cookie = xcb_randr_get_monitors(kwinApp()->x11Connection(), kwinApp()->x11RootWindow(), 1);
+    const UniqueCPtr<xcb_randr_get_monitors_reply_t> monitors(xcb_randr_get_monitors_reply(kwinApp()->x11Connection(), cookie, nullptr));
+    if (!monitors) {
         return nullptr;
     }
 
-    const UniqueCPtr<xcb_xinerama_query_screens_reply_t> screens(xcb_xinerama_query_screens_reply(connection, xcb_xinerama_query_screens(connection), nullptr));
-    if (!screens) {
-        return nullptr;
+    xcb_randr_monitor_info_t *monitorInfo = nullptr;
+    for (auto it = xcb_randr_get_monitors_monitors_iterator(monitors.get()); it.rem; xcb_randr_monitor_info_next(&it)) {
+        const int current = monitors->nMonitors - it.rem;
+        if (current == index) {
+            monitorInfo = it.data;
+            break;
+        }
     }
 
-    const int infoCount = xcb_xinerama_query_screens_screen_info_length(screens.get());
-    if (index < 0 || index >= infoCount) {
+    if (!monitorInfo) {
         return nullptr;
     }
 
-    const xcb_xinerama_screen_info_t *infos = xcb_xinerama_query_screens_screen_info(screens.get());
-    const QRect needle(infos[index].x_org, infos[index].y_org, infos[index].width, infos[index].height);
-
-    for (Output *output : std::as_const(m_outputs)) {
-        if (Xcb::toXNative(output->geometryF()) == needle) {
-            return output;
-        }
-    }
-
-    return nullptr;
+    return findOutput(Xcb::atomName(monitorInfo->name));
 }
 #endif
 
diff --git a/src/workspace.h b/src/workspace.h
index e31f3702486..3112513bc35 100644
--- a/src/workspace.h
+++ b/src/workspace.h
@@ -357,6 +357,7 @@ public:
         DirectionNext
     };
     Output *findOutput(Output *reference, Direction direction, bool wrapAround = false) const;
+    Output *findOutput(const QString &name) const;
     void switchToOutput(Output *output);
 
     QList<Output *> outputs() const;
diff --git a/src/xwayland/drag_x.cpp b/src/xwayland/drag_x.cpp
index 93d0c9d1f62..d61f1beff6b 100644
--- a/src/xwayland/drag_x.cpp
+++ b/src/xwayland/drag_x.cpp
@@ -51,7 +51,7 @@ static QStringList atomToMimeTypes(xcb_atom_t atom)
     } else if (atom == atoms->netscape_url) {
         mimeTypes << QStringLiteral("_NETSCAPE_URL");
     } else {
-        mimeTypes << Selection::atomName(atom);
+        mimeTypes << Xcb::atomName(atom);
     }
     return mimeTypes;
 }
diff --git a/src/xwayland/selection.cpp b/src/xwayland/selection.cpp
index 3f489b1a9a6..4409075332b 100644
--- a/src/xwayland/selection.cpp
+++ b/src/xwayland/selection.cpp
@@ -45,21 +45,6 @@ xcb_atom_t Selection::mimeTypeToAtomLiteral(const QString &mimeType)
     return Xcb::Atom(mimeType.toLatin1(), false, kwinApp()->x11Connection());
 }
 
-QString Selection::atomName(xcb_atom_t atom)
-{
-    xcb_connection_t *xcbConn = kwinApp()->x11Connection();
-    xcb_get_atom_name_cookie_t nameCookie = xcb_get_atom_name(xcbConn, atom);
-    xcb_get_atom_name_reply_t *nameReply = xcb_get_atom_name_reply(xcbConn, nameCookie, nullptr);
-    if (!nameReply) {
-        return QString();
-    }
-
-    const size_t length = xcb_get_atom_name_name_length(nameReply);
-    QString name = QString::fromLatin1(xcb_get_atom_name_name(nameReply), length);
-    free(nameReply);
-    return name;
-}
-
 QStringList Selection::atomToMimeTypes(xcb_atom_t atom)
 {
     QStringList mimeTypes;
@@ -74,7 +59,7 @@ QStringList Selection::atomToMimeTypes(xcb_atom_t atom)
     } else if (atom == atoms->targets || atom == atoms->timestamp) {
         // Ignore known ICCCM internal atoms
     } else {
-        const QString atomNameName = atomName(atom);
+        const QString atomNameName = Xcb::atomName(atom);
         // Ignore other non-mimetype atoms
         if (atomNameName.contains(QLatin1Char('/'))) {
             mimeTypes << atomNameName;
diff --git a/src/xwayland/selection.h b/src/xwayland/selection.h
index 438fc4c6a6b..65a6ca87c09 100644
--- a/src/xwayland/selection.h
+++ b/src/xwayland/selection.h
@@ -49,7 +49,6 @@ public:
     static xcb_atom_t mimeTypeToAtom(const QString &mimeType);
     static xcb_atom_t mimeTypeToAtomLiteral(const QString &mimeType);
     static QStringList atomToMimeTypes(xcb_atom_t atom);
-    static QString atomName(xcb_atom_t atom);
     static void sendSelectionNotify(xcb_selection_request_event_t *event, bool success);
 
     // on selection owner changes by X clients (Xwl -> Wl)
-- 
GitLab

